lib/pymedphys/_dicom/connect/listen.py

```diff
@@ -0,0 +1,207 @@
+    AE,
+    PYNETDICOM_IMPLEMENTATION_UID,
+    PYNETDICOM_IMPLEMENTATION_VERSION,
+    StoragePresentationContexts,
+    evt,
+)
+from pynetdicom.sop_class import (  # pylint: disable=no-name-in-module
+    VerificationSOPClass,
+)
+
+from pymedphys._dicom.connect.base import DicomConnectBase
+
+
+class DicomListener(DicomConnectBase):
+    def __init__(
+        self, storage_directory=tempfile.mkdtemp(), on_released_callback=None, **kwargs
+    ):
+        super(DicomListener, self).__init__(**kwargs)
+
+        # The directory where incoming data will be stored
+        self.storage_directory = storage_directory
+        if not isinstance(self.storage_directory, Path):
+            self.storage_directory = Path(self.storage_directory)
+
+        # Callback when association is released called with path to the directory where
+        # data from that association was stored
+        self.on_released_callback = on_released_callback
+
+        # Keep track of the directory where data for the current association is stored
+        self.association_directory = None
+
+        # The application entity
+        self.ae = None
+
+    def start(self):
+
+        # Initialise the Application Entity
+        self.ae = AE(ae_title=self.ae_title)
+
+        # Add the supported presentation context
+        self.ae.add_supported_context(VerificationSOPClass)
+        for context in StoragePresentationContexts:
+            self.ae.add_supported_context(context.abstract_syntax)
+
+        handlers = [
+            (evt.EVT_C_MOVE, self.on_c_store),
+            (evt.EVT_C_STORE, self.on_c_store),
+            (evt.EVT_C_ECHO, self.on_c_echo),
+            (evt.EVT_ACCEPTED, self.on_association_accepted),
+            (evt.EVT_RELEASED, self.on_association_released),
+        ]
+
+        # Start listening for incoming association requests
+        self.ae.start_server((self.host, self.port), evt_handlers=handlers, block=False)
+
+    def stop(self):
+
+        if self.ae:
+            self.ae.shutdown()
+
+    def on_c_echo(self, _):
+        """Respond to a C-ECHO service request.
+        """
+        logging.debug("C-ECHO!")
+        return 0x0000
+
+    def on_association_accepted(self, _):
+        self.association_directory = None
+
+    def on_association_released(self, _):
+
+        if self.on_released_callback:
+            self.on_released_callback(self.association_directory)
+
+    def on_c_store(self, event):
+
+        dataset = event.dataset
+
+        # TODO: Should these be place in a config file?
+        mode_prefixes = {
+            "CT Image Storage": "CT",
+            "Enhanced CT Image Storage": "CTE",
+            "MR Image Storage": "MR",
+            "Enhanced MR Image Storage": "MRE",
+            "Positron Emission Tomography Image Storage": "PT",
+            "Enhanced PET Image Storage": "PTE",
+            "RT Image Storage": "RI",
+            "RT Dose Storage": "RD",
+            "RT Plan Storage": "RP",
```

Comment:
RT Ion Plan?
RT Beams Treatment Record (this is used in conventional, so it is relevant to typical departments)
RT Ion Beams Treatment Record